File: app/src/main/java/moe/feng/common/eventshelper/sample/SampleActivity.java
package moe.feng.common.eventshelper.sample;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.Nullable;

import moe.feng.common.eventshelper.EventsHelper;

public class SampleActivity extends Activity {

    private static final String TAG = SampleActivity.class.getSimpleName();

    private TestListener mListenerA = (time) -> Log.d(TAG, "ListenerA: " + time + " thread: " + Thread.currentThread());
    private TestAsyncListener mListenerB = (time) -> Log.d(TAG, "ListenerB: " + time + " thread: " + Thread.currentThread());
    private TestAsyncListener mListenerC = (time) -> Log.d(TAG, "ListenerC: " + time + " thread: " + Thread.currentThread());

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sample);

        EventsHelper.init(this);

        findViewById(R.id.send_event_button).setOnClickListener(v -> {
            Log.d(TAG, "Now calling A TestListener#onButtonClick");
            EventsHelper.of(TestListener.class, "A").onButtonClick(System.currentTimeMillis());
            Log.d(TAG, "Now calling B TestListener#onButtonClick");
            EventsHelper.of(TestAsyncListener.class, "B").onButtonClick(System.currentTimeMillis());
            Log.d(TAG, "Now calling C TestListener#onButtonClick");
            EventsHelper.of(TestAsyncListener.class, "C").onButtonClick(System.currentTimeMillis());
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        EventsHelper.registerListener(mListenerA, "A");
        EventsHelper.registerListener(mListenerB, "B");
        EventsHelper.registerListener(mListenerC, "C");
    }

    @Override
    protected void onStop() {
        super.onStop();
        EventsHelper.unregisterListeners(mListenerA, mListenerB, mListenerC);
    }

}

File: app/src/main/java/moe/feng/common/eventshelper/sample/SampleActivity.kt
package moe.feng.common.eventshelper.sample

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.view.View

import moe.feng.common.eventshelper.EventsHelper
import moe.feng.common.eventshelper.listeners
import moe.feng.common.eventshelper.of

class SampleActivity : Activity() {

    companion object {

        private val TAG = SampleActivity::class.java.simpleName

    }

    private val mListenerA = object : TestListener {
        override fun onButtonClick(time: Long) {
            Log.d(TAG, "ListenerA: " + time + " thread: " + Thread.currentThread())
        }

        override fun ignoredMethod() {
            throw RuntimeException("This method shouldn't be called by EventsHelper")
        }
    }
    private val mListenerB = TestAsyncListener { time ->
        Log.d(TAG, "ListenerB: " + time + " thread: " + Thread.currentThread())
    }
    private val mListenerC = TestAsyncListener { time ->
        Log.d(TAG, "ListenerC: " + time + " thread: " + Thread.currentThread())
    }

    private lateinit var mEventsHelper: EventsHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sample)

        mEventsHelper = EventsHelper.getInstance(this)

        findViewById<View>(R.id.send_event_button).setOnClickListener {
            Log.d(TAG, "Now calling TestListener#onButtonClick")
            mEventsHelper.of<TestListener>().onButtonClick(System.currentTimeMillis())
            Log.d(TAG, "Now calling B TestListener#onButtonClick")
            mEventsHelper.of<TestAsyncListener>("B").onButtonClick(System.currentTimeMillis())
            Log.d(TAG, "Now calling C TestListener#onButtonClick")
            mEventsHelper.of<TestAsyncListener>("C").onButtonClick(System.currentTimeMillis())

            Log.d(TAG, "Now calling TestListener#ignoredMethod with try-catch")
            try {
                mEventsHelper.of<TestListener>().ignoredMethod()
            } catch (e: Exception) {
                Log.e(TAG, "ignoredMethod of helper throws a exception", e)
            }
        }
    }

    override fun onStart() {
        super.onStart()
        mEventsHelper.listeners += mListenerA to "A"
        mEventsHelper.listeners += listOf(
                mListenerB to "B",
                mListenerC to "C"
        )
        // Java style
        // mEventsHelper.registerListener(mListenerA, "A")
        // mEventsHelper.registerListener(mListenerB, "B")
        // mEventsHelper.registerListener(mListenerC, "C")
    }

    override fun onStop() {
        super.onStop()
        mEventsHelper.listeners -= listOf(mListenerA, mListenerB, mListenerC)
        // Java style
        // mEventsHelper.unregisterListeners(mListenerA, mListenerB, mListenerC)
    }

}
